man/bppseqman.1
.\" SPDX-FileCopyrightText: The Bio++ Development Group
.\"
.\" SPDX-License-Identifier: CECILL-2.1

.TH BPPSEQMAN 1 LOCAL

.SH NAME

bppseqman - Bio++ sequence manipulation and in silico molecular biology with Bio++

.SH SYNOPSIS

.B bppseqman [options]

.SH AVAILABILITY

All UNIX flavors

.SH DESCRIPTION

bppseqman performs several in silico molecular biology operations like transcription, translation or traduction. It also allows conversion between various file formats, and bioinformatics tasks like removing of gaps, unknown characters or stop codons.

.SH OPTIONS

You should refer to 'info bppsuite' or to the online manual of bppsuite for a complete list of available options.

.TP 5

--noninteractive

generates output for redirection in a file.

.TP

param=file

reads a file for loading options

.SH AUTHOR

Bio++ Development Team.


ChangeLog
19/03/18 -*- Version 2.4.0 -*-

19/02/18 Julien Dutheil
* Bug fixed in bppPopStats: dN and dS were inverted!

10/12/17 -*- Version 2.3.2 -*-

06/11/17 Julien Dutheil
* Added estimation of kappa to bppPopStats + bugs fixed

06/06/17 -*- Version 2.3.1 -*- 

10/05/17 -*- Version 2.3.0 -*- 

21/12/16 Julien Dutheil
* BppPhySamp migrated to its own package.

17/12/16 Julien Dutheil
* Improved bppPopStats, outgroup analyses, MK table and output to file.

06/10/16 Julien Dutheil
* bppSeqGen now takes Hudson's MS ARG files as input.
* new input.tree.scale option to adjust branch lengths.

22/05/16 Julien Dutheil
* bpppopstats now computes codon statistics per site (e.g. PiN, PiS)
* Added example for bpppopstats.

29/06/15 Julien Dutheil
* New program bppPopStats

28/01/15 Julien Dutheil
* PhySamp now outputs the correct tree!

25/11/14 Julien Dutheil
* bppalnscore output contiguous segments in mase selection.
* bppalnscore site selection works with codon alphabets.

24/11/14 Julien Dutheil
* bppSeqMan can now read alignments and perform site selection priori to any analysis.

28/09/14 -*- Version 2.2.0 -*- 

16/09/14 Julien Dutheil
* bppSeqGen supports generic characters as ancestral states in info file (bug #87).

12/08/14 Julien Dutheil
* bppSeqGen supports simulation under Markov-Modulated models.

06/06/14 Julien Dutheil
* bppPhySamp now output trees as well.

17/04/14 Julien Dutheil
* bppSeqMan: Translate uses global genetic_code argument instead of local 'code' argument.
* All programs use the --seed argument to set random seed for reproduceable results.

08/03/13 -*- Version 0.8.0 -*-

22/01/13 Julien Dutheil
* Compiles with -Wconversion

14/01/13 Mathieu Groussin
* Added COaLA model for proteins in bppML and bppSeqGen.

23/11/12 Laurent Guéguen
* New bppMixedLikelihood program

08/02/12 -*- Version 0.7.0 -*-

21/12/11 Julien Dutheil
* New bppAlnScore program
* Improved bppSeqMan: protein translation now works + more stop codons removal options.

17/06/11 Julien Dutheil
* Improved bppancestor, also allows to compute frequencies for leaves.

11/06/11 -*- Version 0.6.2 -*-
* Small bug fixed + new packages.

03/06/10 -*- Version 0.6.0 -*-
* Compatibility with Bio++ 2.0.0.

03/06/10 Julien Dutheil
* New method in bppPhySamp + new, faster, implementation.

25/03/10 -*- Version 0.5.0 -*-

22/08/09 Julien Dutheil
* Added CMake configuration files, with twick for building the manual (assumes
		texinfo is installed).

22/07/09 Julien Dutheil
* Added program bppTreeDraw.

30/06/09 Julien Dutheil
* BppAncestor: now output expected ancestral states frequencies.
* BppAncestor: new method to compute EB estimates of ancestral states frequencies.

10/06/09 -*- Version 0.4.0 -*-

22/05/09 Julien Dutheil
* Keyval syntax for sequence format description.
* Documentation and example update.
* SeqMan now uses the SequenceApplicationTools method.

15/05/09 Julien Dutheil
* Compatibility update with the libs, new keyval syntax for model
specification.
* Documentation up to date.

02/02/09 Julien Dutheil
* bppConsense now outputs the number of leaves of the input tree.
* New progress bar in bppConsense.

08/01/09 Julien Dutheil
* Added texinfo documentation.
* More bootstrap options in BppDist (the same as in BppML).

25/09/08 -*- Version 0.3.1 -*-

07/12/08 Bastien Boussau & Julien Dutheil
* Bug removed in non-homogeneous models with covarions.

09/10/08 Julien Dutheil
* New action 'invert' in bppSeqMan.

25/09/08 -*- Version 0.3.0 -*-

20/09/08 Julien Dutheil
* New program bppAncestor to reconstruct acestral states and rates.
* New +F models from PhylLib. 

02/07/08 Laurent Guegen
* Bug fixed in bppML: frequencies can be computed from the data with
non-homogeneous models.

19/06/08 Julien Dutheil
* Shell variables are now 'global' in option files.

24/04/08 -*- Version 0.2.3 -*-
* This version requires the CVS version of Bio++ to be built.

03/03/08 Julien Dutheil
* Bug fixed in keep_complete method of BppSeqMan.

27/03/08 Julien Dutheil
* Bug fixed when printing parameters of non-homogeneous models in bppML.

22/01/08 -*- Version 0.2.2 -*-

28/02/08 Celine Scronavacca
* Bug fixed in ReRoot.

22/01/08 -*- Version 0.2.1 -*-

06/02/08 Celine Scornavacca
* Bug fixed in ReRoot + correction of version number in bppSeqGen.

22/01/08 -*- Version 0.2.0 -*-

13/01/08 Julien Dutheil
* More options in bppml: possibility to chose between simple and double
recursion and data compression method.
* Simple recursion method improved.

11/01/08 Julien Dutheil
* Compatibility update

bppSuite/bppSeqMan.cpp
// SPDX-FileCopyrightText: The Bio++ Development Group
//
// SPDX-License-Identifier: CECILL-2.1

// From the STL:
#include <iostream>
#include <fstream>
#include <iomanip>

using namespace std;

// From bpp-core:
#include <Bpp/Version.h>
#include <Bpp/App/BppApplication.h>
#include <Bpp/App/ApplicationTools.h>
#include <Bpp/Text/KeyvalTools.h>

// From bpp-seq:
#include <Bpp/Seq/SiteTools.h>
#include <Bpp/Seq/CodonSiteTools.h>
#include <Bpp/Seq/Alphabet/Alphabet.h>
#include <Bpp/Seq/Alphabet/AlphabetTools.h>
#include <Bpp/Seq/Container/VectorSiteContainer.h>
#include <Bpp/Seq/Container/SiteContainerTools.h>
#include <Bpp/Seq/Container/SequenceContainerTools.h>
#include <Bpp/Seq/App/SequenceApplicationTools.h>
#include <Bpp/Seq/SequenceTools.h>

// From bpp-phyl:
#include <Bpp/Phyl/Tree/Tree.h>
#include <Bpp/Phyl/App/PhylogeneticsApplicationTools.h>

using namespace bpp;

void help()
{
  (*ApplicationTools::message << "__________________________________________________________________________").endLine();
  (*ApplicationTools::message << "bppseqman parameter1_name=parameter1_value").endLine();
  (*ApplicationTools::message << "      parameter2_name=parameter2_value ... param=option_file").endLine();
  (*ApplicationTools::message).endLine();
  (*ApplicationTools::message << "  Refer to the Bio++ Program Suite Manual for a list of available options.").endLine();
  (*ApplicationTools::message << "__________________________________________________________________________").endLine();
}

int main(int args, char** argv)
{
  cout << "******************************************************************" << endl;
  cout << "*           Bio++ Sequence Manipulator, version " << BPP_VERSION << ".           *" << endl;
  cout << "* Author: J. Dutheil                        Last Modif. " << BPP_REL_DATE << " *" << endl;
  cout << "******************************************************************" << endl;
  cout << endl;

  if (args == 1)
  {
    help();
    return 0;
  }

  try
  {
    BppApplication bppseqman(args, argv, "BppSeqMan");
    bppseqman.startTimer();

    // Get alphabet
    shared_ptr<Alphabet> alphabet = SequenceApplicationTools::getAlphabet(bppseqman.getParams(), "", false, true, true);

    shared_ptr<CodonAlphabet> codonAlphabet = dynamic_pointer_cast<CodonAlphabet>(alphabet);

    // Get sequences:
    bool aligned = ApplicationTools::getBooleanParameter("input.alignment", bppseqman.getParams(), false, "", true, 1);
    shared_ptr<SequenceContainerInterface> sequences = 0;

    if (aligned)
    {
      shared_ptr<VectorSiteContainer> allSites = SequenceApplicationTools::getSiteContainer(alphabet, bppseqman.getParams());
      sequences = SequenceApplicationTools::getSitesToAnalyse(*allSites, bppseqman.getParams(), "", true, false);
    }
    else
    {
      sequences = SequenceApplicationTools::getSequenceContainer(alphabet, bppseqman.getParams(), "", true, true);
    }

    ApplicationTools::displayResult("Number of sequences", sequences->getNumberOfSequences());

    // Perform manipulations

    vector<string> actions = ApplicationTools::getVectorParameter<string>("sequence.manip", bppseqman.getParams(), ',', "", "", false, 1);


    for (size_t a = 0; a < actions.size(); a++)
    {
      string cmdName;
      map<string, string> cmdArgs;
      KeyvalTools::parseProcedure(actions[a], cmdName, cmdArgs);
      ApplicationTools::displayResult("Performing action", cmdName);

      // +-----------------+
      // | Complementation |
      // +-----------------+
      if (cmdName == "Complement")
      {
        shared_ptr<SequenceContainerInterface> sc = 0;
        if (aligned)
          sc = make_shared<VectorSiteContainer>(sequences->getAlphabet());
        else
          sc = make_shared<VectorSequenceContainer>(sequences->getAlphabet());
        for (size_t i = 0; i < sequences->getNumberOfSequences(); i++)
        {
          auto seq = SequenceTools::getComplement(sequences->sequence(i));
          auto name = sequences->sequenceKey(i);
          sc->addSequence(name, seq);
        }
        sequences = sc;
      }
      // +------------------------+
      // | (Reverse)Transcription |
      // +------------------------+
      else if (cmdName == "Transcript")
      {
        if (sequences->alphabet().getAlphabetType() == AlphabetTools::DNA_ALPHABET->getAlphabetType())
        {
          shared_ptr<SequenceContainerInterface> sc = 0;
          if (aligned)
            sc = make_shared<VectorSiteContainer>(AlphabetTools::RNA_ALPHABET);
          else
            sc = make_shared<VectorSequenceContainer>(AlphabetTools::RNA_ALPHABET);
          for (unsigned int i = 0; i < sequences->getNumberOfSequences(); i++)
          {
            auto seq = SequenceTools::transcript(sequences->sequence(i));
            auto name = sequences->sequenceKey(i);
            sc->addSequence(name, seq);
          }
          sequences = sc;
        }
        else if (sequences->alphabet().getAlphabetType() == AlphabetTools::RNA_ALPHABET->getAlphabetType())
        {
          shared_ptr<SequenceContainerInterface> sc = 0;
          if (aligned)
            sc = make_shared<VectorSiteContainer>(AlphabetTools::DNA_ALPHABET);
          else
            sc = make_shared<VectorSequenceContainer>(AlphabetTools::DNA_ALPHABET);
          for (unsigned int i = 0; i < sequences->getNumberOfSequences(); i++)
          {
            auto seq = SequenceTools::reverseTranscript(sequences->sequence(i));
            auto name = sequences->sequenceKey(i);
            sc->addSequence(name, seq);
          }
          sequences = sc;
        }
        else
          throw Exception("Transcription error: input alphabet must be of type 'nucleic'.");
      }
      // +-------------------------------+
      // | Switching nucleotide alphabet |
      // +-------------------------------+
      else if (cmdName == "Switch")
      {
        std::shared_ptr<const Alphabet> alpha = 0;
        if (sequences->alphabet().getAlphabetType() == AlphabetTools::DNA_ALPHABET->getAlphabetType())
        {
          alpha = AlphabetTools::RNA_ALPHABET;
        }
        else if (sequences->alphabet().getAlphabetType() == AlphabetTools::RNA_ALPHABET->getAlphabetType())
        {
          alpha = AlphabetTools::DNA_ALPHABET;
        }
        else
          throw Exception("Cannot switch alphabet type, alphabet is not of type 'nucleic'.");

        shared_ptr<SequenceContainerInterface> sc = 0;
        if (aligned)
          sc = make_shared<VectorSiteContainer>(alpha);
        else
          sc = make_shared<VectorSequenceContainer>(alpha);

        for (size_t i = 0; i < sequences->getNumberOfSequences(); i++)
        {
          auto old = sequences->sequence(i);
          vector<int> content(old.size());
          for (size_t j = 0; j < old.size(); ++j)
          {
            content[j] = old[j];
          }
          auto seq = make_unique<Sequence>(old.getName(), content, old.getComments(), alpha);
          auto name = sequences->sequenceKey(i);
          sc->addSequence(name, seq);
        }
        sequences = sc;
      }
      // +-------------+
      // | Translation |
      // +-------------+
      else if (cmdName == "Translate")
      {
        if (!AlphabetTools::isCodonAlphabet(sequences->alphabet()))
          throw Exception("Error in translation: alphabet is not of type 'codon'.");

        string codeDesc = ApplicationTools::getStringParameter("genetic_code", bppseqman.getParams(), "Standard", "", true, 1);

        auto gCode = SequenceApplicationTools::getGeneticCode(codonAlphabet->getNucleicAlphabet(), codeDesc);

        ApplicationTools::displayResult("Genetic Code", codeDesc);

        shared_ptr<SequenceContainerInterface> sc = 0;
        if (aligned)
          sc = make_shared<VectorSiteContainer>(AlphabetTools::PROTEIN_ALPHABET);
        else
          sc = make_shared<VectorSequenceContainer>(AlphabetTools::PROTEIN_ALPHABET);

        for (size_t i = 0; i < sequences->getNumberOfSequences(); ++i)
        {
          auto seq = gCode->translate(sequences->sequence(i));
          auto name = sequences->sequenceKey(i);
          sc->addSequence(name, seq);
        }
        sequences = sc;
      }
      // +-------------+
      // | Remove gaps |
      // +-------------+
      else if (cmdName == "RemoveGaps")
      {
        auto sc = make_shared<VectorSequenceContainer>(sequences->getAlphabet());

        for (size_t i = 0; i < sequences->getNumberOfSequences(); i++)
        {
          unique_ptr<Sequence> seq(sequences->sequence(i).clone());
          SequenceTools::removeGaps(*seq);
          auto name = sequences->sequenceKey(i);
          sc->addSequence(name, seq);
        }
        sequences = sc;
        aligned = false;
      }
      // +---------------------------+
      // | Change gaps to unresolved |
      // +---------------------------+
      else if (cmdName == "GapToUnknown")
      {
        shared_ptr<SequenceContainerInterface> sc = 0;
        if (aligned)
          sc = make_shared<VectorSiteContainer>(sequences->getAlphabet());
        else
          sc = make_shared<VectorSequenceContainer>(sequences->getAlphabet());

        for (size_t i = 0; i < sequences->getNumberOfSequences(); i++)
        {
          auto seq = make_unique<Sequence>(sequences->sequence(i));
          SymbolListTools::changeGapsToUnknownCharacters(*seq);
          auto name = sequences->sequenceKey(i);
          sc->addSequence(name, seq);
        }
        sequences = sc;
      }
      // +---------------------------+
      // | Change unresolved to gaps |
      // +---------------------------+
      else if (cmdName == "UnknownToGap")
      {
        shared_ptr<SequenceContainerInterface> sc = 0;
        if (aligned)
          sc = make_shared<VectorSiteContainer>(sequences->getAlphabet());
        else
          sc = make_shared<VectorSequenceContainer>(sequences->getAlphabet());

        for (size_t i = 0; i < sequences->getNumberOfSequences(); i++)
        {
          auto seq = make_unique<Sequence>(sequences->sequence(i));
          SymbolListTools::changeUnresolvedCharactersToGaps(*seq);
          auto name = sequences->sequenceKey(i);
          sc->addSequence(name, seq);
        }
        sequences = sc;
      }

      // +--------------+
      // | Remove stops |
      // +--------------+
      else if (cmdName == "RemoveStops")
      {
        if (!codonAlphabet)
          throw Exception("RemoveStops: requires a codon alphabet.");

        string codeDesc = ApplicationTools::getStringParameter("genetic_code", bppseqman.getParams(), "Standard", "", true, 1);
        ApplicationTools::displayResult("Genetic Code", codeDesc);
        auto gCode = SequenceApplicationTools::getGeneticCode(codonAlphabet->getNucleicAlphabet(), codeDesc);

        auto sites = dynamic_pointer_cast<SiteContainerInterface>(sequences);
        if (!sites)
        {
          auto sc = make_shared<VectorSequenceContainer>(sequences->getAlphabet());
          for (size_t i = 0; i < sequences->getNumberOfSequences(); ++i)
          {
            unique_ptr<Sequence> seq(sequences->sequence(i).clone());
            SequenceTools::removeStops(*seq, *gCode);
            auto name = sequences->sequenceKey(i);
            sc->addSequence(name, seq);
          }
          sequences = sc;
        }
        else
        {
          auto sc = make_unique<VectorSiteContainer>(sequences->getAlphabet());
          for (size_t i = 0; i < sequences->getNumberOfSequences(); ++i)
          {
            unique_ptr<Sequence> seq(sequences->sequence(i).clone());
            SequenceTools::replaceStopsWithGaps(*seq, *gCode);
            auto name = sequences->sequenceKey(i);
            sc->addSequence(name, seq);
          }
          sequences.reset(sc.release());
        }
      }

      // +--------------+
      // | Remove stops |
      // +--------------+
      else if (cmdName == "RemoveColumnsWithStops")
      {
        auto sites = dynamic_pointer_cast<SiteContainerInterface>(sequences);
        if (!sites)
        {
          throw Exception("'RemoveColumnsWithStops' can only be used on alignment. You may consider using the 'CoerceToAlignment' command.");
        }
        if (!codonAlphabet)
          throw Exception("RemoveColumnsWithStops: requires a codon alphabet.");
        string codeDesc = ApplicationTools::getStringParameter("genetic_code", bppseqman.getParams(), "Standard", "", true, 1);

        ApplicationTools::displayResult("Genetic Code", codeDesc);
        auto gCode = SequenceApplicationTools::getGeneticCode(codonAlphabet->getNucleicAlphabet(), codeDesc);

        for (size_t i = sites->getNumberOfSites(); i > 0; i--)
        {
          if (CodonSiteTools::hasStop(sites->site(i - 1), *gCode))
            sites->deleteSite(i - 1);
        }
      }

      // +---------+
      // | Get CDS |
      // +---------+
      else if (cmdName == "GetCDS")
      {
        if (!codonAlphabet)
          throw Exception("GetCDS: requires a codon alphabet.");
        string codeDesc = ApplicationTools::getStringParameter("genetic_code", bppseqman.getParams(), "Standard", "", true, 1);
        ApplicationTools::displayResult("Genetic Code", codeDesc);
        auto gCode = SequenceApplicationTools::getGeneticCode(codonAlphabet->getNucleicAlphabet(), codeDesc);

        shared_ptr<SequenceContainerInterface> sc = 0;
        if (aligned)
          sc = make_shared<VectorSiteContainer>(sequences->getAlphabet());
        else
          sc = make_shared<VectorSequenceContainer>(sequences->getAlphabet());

        for (size_t i = 0; i < sequences->getNumberOfSequences(); ++i)
        {
          auto seq = unique_ptr<Sequence>(sequences->sequence(i).clone());
          size_t len = seq->size();
          SequenceTools::getCDS(*seq, *gCode, false, true, true, false);
          if (aligned)
          {
            for (size_t c = seq->size(); c < len; ++c)
            {
              seq->addElement(seq->alphabet().getGapCharacterCode());
            }
          }
          auto name = sequences->sequenceKey(i);
          sc->addSequence(name, seq);
        }
        sequences = sc;
      }

      // +---------------------+
      // | Coerce to alignment |
      // +---------------------+
      else if (actions[a] == "CoerceToAlignment")
      {
        shared_ptr<SiteContainerInterface> sites = dynamic_pointer_cast<SiteContainerInterface>(sequences);
        if (!sites)
        {
          sites = make_shared<VectorSiteContainer>(*sequences);
          sequences = sites;
        }
        aligned = true;
      }
      // +--------------------------+
      // | Resolve dotted alignment |
      // +--------------------------+
      else if (actions[a] == "ResolvedDotted")
      {
        shared_ptr<SiteContainerInterface> sites = dynamic_pointer_cast<SiteContainerInterface>(sequences);
        if (!sites)
        {
          throw Exception("'ResolvedDotted' can only be used on alignment. You may consider using the 'CoerceToAlignment' command.");
        }

        shared_ptr<const Alphabet> alpha = 0;
        string alphastr = ApplicationTools::getStringParameter("alphabet", cmdArgs, "DNA", "", false, 1);
        if (alphastr == "DNA")
          alpha = AlphabetTools::DNA_ALPHABET;
        else if (alphastr == "RNA")
          alpha = AlphabetTools::RNA_ALPHABET;
        else if (alphastr == "Protein")
          alpha = AlphabetTools::PROTEIN_ALPHABET;
        else
          throw Exception("Resolved alphabet must be one of [DNA|RNA|Protein] for solving dotted alignment.");

        shared_ptr<SiteContainerInterface> resolvedCont = SiteContainerTools::resolveDottedAlignment(*sites, alpha);
        sequences = resolvedCont;
      }
      // +---------------------+
      // | Keep complete sites |
      // +---------------------+
      else if (cmdName == "KeepComplete")
      {
        auto sites = dynamic_pointer_cast<SiteContainerInterface>(sequences);
        if (!sites)
        {
          throw Exception("'KeepComplete' can only be used on alignment. You may consider using the 'CoerceToAlignment' command.");
        }

        if (ApplicationTools::parameterExists("maxGapAllowed", cmdArgs))
          throw Exception("Argument maxGapAllowed is deprecated and has been replaced by max_gap_allowed.");
        string maxGapOption = ApplicationTools::getStringParameter("max_gap_allowed", cmdArgs, "100%", "", false, 1);
        if (maxGapOption[maxGapOption.size() - 1] == '%')
        {
          double gapFreq = TextTools::toDouble(maxGapOption.substr(0, maxGapOption.size() - 1)) / 100.;
          for (size_t i = sites->getNumberOfSites(); i > 0; i--)
          {
            map<int, double> freqs;
            SiteTools::getFrequencies(sites->site(i - 1), freqs);
            if (freqs[-1] > gapFreq)
              sites->deleteSite(i - 1);
          }
        }
        else
        {
          size_t gapNum = TextTools::to<size_t>(maxGapOption);
          for (size_t i = sites->getNumberOfSites(); i > 0; i--)
          {
            map<int, size_t> counts;
            SiteTools::getCounts(sites->site(i - 1), counts);
            counts[-1]; // Needed in case this entry does not exist in the map. This will set it to 0.
            if (counts[-1] > gapNum)
              sites->deleteSite(i - 1);
          }
        }
      }
      // +----------------+
      // | Get consensus |
      // +---------------+
      else if (cmdName == "Consensus")
      {
        auto sites = dynamic_pointer_cast<SiteContainerInterface>(sequences);
        if (!sites)
        {
          throw Exception("'Consensus' can only be used on alignment. You may consider using the 'CoerceToAlignment' command.");
        }

        bool ignoreGaps = ApplicationTools::getBooleanParameter("ignore_gaps", cmdArgs, true, "", false, 1);
        bool resolveUnknown = ApplicationTools::getBooleanParameter("resolve_unknown", cmdArgs, true, "", false, 1);

        auto consensusSeq = SiteContainerTools::getConsensus(*sites, "consensus", ignoreGaps, resolveUnknown);
        sites->addSequence("consensus", consensusSeq);
      }
      // +-----------------+
      // | Invert sequence |
      // +-----------------+
      else if (cmdName == "Invert")
      {
        shared_ptr<SequenceContainerInterface> sc = 0;
        if (aligned)
          sc = make_shared<VectorSiteContainer>(sequences->getAlphabet());
        else
          sc = make_shared<VectorSequenceContainer>(sequences->getAlphabet());

        for (size_t i = 0; i < sequences->getNumberOfSequences(); i++)
        {
          auto seq = make_unique<Sequence>(*SequenceTools::getInvert(sequences->sequence(i)).release());
          auto name = sequences->sequenceKey(i);
          sc->addSequence(name, seq);
        }
        sequences = sc;
      }
      // +------------------+
      // | GetCodonPosition |
      // +------------------+
      else if (cmdName == "GetCodonPosition")
      {
        unsigned int pos = ApplicationTools::getParameter<unsigned int>("position", cmdArgs, 3, "", false, 1);

        shared_ptr<SequenceContainerInterface> sc = SequenceContainerTools::getCodonPosition(*sequences, pos - 1);
        //      auto sc = shared_ptr<SequenceContainerInterface>(gp);

        if (aligned)
        {
          sequences = make_shared<VectorSiteContainer>(*sc);
        }
        else
        {
          sequences = sc;
        }
      }
      // +-----------------+
      // | FilterFromTree |
      // +-----------------+
      else if (cmdName == "FilterFromTree")
      {
        unique_ptr<Tree> tree(PhylogeneticsApplicationTools::getTree(cmdArgs, ""));
        vector<string> names = tree->getLeavesNames();

        shared_ptr<SequenceContainerInterface> reorderedSequences = 0;

        if (aligned)
        {
          reorderedSequences = make_shared<VectorSiteContainer>(sequences->getAlphabet());
        }
        else
        {
          reorderedSequences = make_shared<VectorSequenceContainer>(sequences->getAlphabet());
        }

        for (size_t i = 0; i < names.size(); ++i)
        {
          auto seq2 = unique_ptr<Sequence>(sequences->sequence(names[i]).clone());
          reorderedSequences->addSequence(names[i], seq2);
        }
        sequences = reorderedSequences;
      }
      // +----------------------+
      // | RemoveEmptySequences |
      // +----------------------+
      else if (cmdName == "RemoveEmptySequences")
      {
        shared_ptr<SequenceContainerInterface> sc = 0;

        if (aligned)
          sc = make_shared<VectorSiteContainer>(sequences->getAlphabet());
        else
          sc = make_shared<VectorSequenceContainer>(sequences->getAlphabet());
        for (size_t i = 0; i < sequences->getNumberOfSequences(); ++i)
        {
          if (SequenceTools::getNumberOfSites(sequences->sequence(i)) != 0)
          {
            auto name = sequences->sequenceKey(i);
            auto seq2 = unique_ptr<Sequence>(sequences->sequence(i).clone());
            sc->addSequence(name, seq2);
          }
        }
        sequences = sc;
      }

      else
        throw Exception("Unknown action: " + cmdName);
    }

    // Write sequences
    ApplicationTools::displayBooleanResult("Final sequences are aligned", aligned);
    if (aligned)
    {
      auto sites = dynamic_pointer_cast<SiteContainerInterface>(sequences);
      SequenceApplicationTools::writeAlignmentFile(*sites, bppseqman.getParams(), "", true, 1);

      // Write index
      string indexPath = ApplicationTools::getAFilePath("output.index.file",  bppseqman.getParams(), false, false);
      if (indexPath != "none" && indexPath != "None")
      {
        ofstream indexFile(indexPath, ios::out);
        indexFile << "# SGED index file version 1.00" << endl;
        indexFile << "# SGED index start" << endl;
        indexFile << "AlnPos,OrigPos" << endl;
        for (size_t i = 0; i < sites->getNumberOfSites(); ++i)
        {
          auto& site = sites->site(i);
          indexFile << (i + 1) << "," << site.getCoordinate() << endl;
        }
        indexFile.close();
      }
    }
    else
    {
      SequenceApplicationTools::writeSequenceFile(*sequences, bppseqman.getParams(), "", true, 1);
    }

    bppseqman.done();
  }
  catch (exception& e)
  {
    cout << e.what() << endl;
    return 1;
  }

  return 0;
}
